lock-api/src/main/java/com/palantir/lock/client/MultiClientBatchingIdentifiedAtlasDbTransactionStarter.java

```diff
@@ -0,0 +1,319 @@
+                        .map(response -> response.immutableTimestamp().getLock())
+                        .collect(Collectors.toSet()),
+                multiClientRequestManager.getLockCleanupService(namespace)));
+    }
+
+    private static Map<Namespace, ResponseHandler> getResponseHandlers(
+            List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>> batch) {
+        Map<Namespace, Queue<SettableResponse>> namespaceWisePendingFutures = new HashMap<>();
+
+        for (BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>> element : batch) {
+            Namespace namespace = element.argument().namespace();
+            namespaceWisePendingFutures
+                    .computeIfAbsent(namespace, _u -> new LinkedList<>())
+                    .add(SettableResponse.of(element.argument().params().numTransactions(), element.result()));
+        }
+
+        return KeyedStream.stream(namespaceWisePendingFutures)
+                .map(ResponseHandler::new)
+                .collectToMap();
+    }
+
+    @VisibleForTesting
+    static Map<Namespace, RequestParams> getNamespaceWiseRequestParams(
+            List<BatchElement<NamespaceAndRequestParams, List<StartIdentifiedAtlasDbTransactionResponse>>> batch) {
+        return batch.stream()
+                .map(BatchElement::argument)
+                .collect(Collectors.toMap(
+                        NamespaceAndRequestParams::namespace,
+                        NamespaceAndRequestParams::params,
+                        RequestParams::coalesce));
+    }
+
+    private static Map<Namespace, List<StartIdentifiedAtlasDbTransactionResponse>> getStartTransactionResponses(
+            Map<Namespace, RequestParams> originalRequestMap,
+            InternalMultiClientConjureTimelockService delegate,
+            UUID requestorId) {
+
+        Map<Namespace, ConjureStartTransactionsRequest> namespaceWiseRequests =
+                getNamespaceWiseRequests(originalRequestMap, requestorId);
+
+        Map<Namespace, ConjureStartTransactionsResponse> responseMap = getResponseMap(delegate, namespaceWiseRequests);
+
+        return KeyedStream.stream(responseMap)
+                .mapEntries((namespace, response) -> {
+                    originalRequestMap
+                            .get(namespace)
+                            .cache()
+                            .processStartTransactionsUpdate(
+                                    response.getTimestamps().stream().boxed().collect(Collectors.toSet()),
+                                    response.getLockWatchUpdate());
+                    LockWatchLogUtility.logTransactionEvents(
+                            fromConjure(namespaceWiseRequests.get(namespace).getLastKnownVersion()),
+                            response.getLockWatchUpdate());
+                    return Maps.immutableEntry(namespace, TransactionStarterHelper.split(response));
```

Comment:
Hmm, there may be some deduplication we can do with `BatchingIdentifiedAtlasDbTransactionStarter`, at least once we've got the cache. Give this a try, though with this one I'm aware there may be unexpected difficulties 🔥